lib/offsite_payments/integrations/latipay.rb

```diff
@@ -0,0 +1,247 @@
+module OffsitePayments
+  module Integrations
+    module Latipay
+
+      def self.notification(post, options = {})
+        Notification.new(post, options)
+      end
+
+      def self.return(query_string, options = {})
+        Return.new(query_string, options)
+      end
+
+      class Interface
+        include ActiveUtils::PostsData # ssl_get/post
+
+        def self.base_url
+          "https://api.latipay.net/v2"
+        end
+
+        def initialize(api_key, user_id)
+          @api_key = api_key
+          @user_id = user_id
+        end
+
+        def sign(fields)
+          message = fields.compact.sort.map{ |k,v| "#{k.to_s}=#{v}" }.join('&').concat(@api_key)
+          OpenSSL::HMAC.hexdigest(OpenSSL::Digest::SHA256.new, @api_key, message)
+        end
+  
+        def verify_signature(message, signature)
+          signature == OpenSSL::HMAC.hexdigest(OpenSSL::Digest::SHA256.new, @api_key, message)
+        end
+
+        private
+
+        def standard_headers
+          {
+            'Content-Type' => 'application/json'
+          }
+        end
+
+        def parse_response(raw_response)
+          JSON.parse(raw_response)
+        end
+
+        class RequestError < StandardError
+          attr_reader :exception, :message, :response_code
+
+          def initialize(response)
+            @response      = response
+            @response_code = @response['code']
+            @success       = @response_code == '0'
+            @message       = @response['message']
+          end
+
+          # Latipay doesn't provide error codes for each interface, this is the only list for all the errors.
+          ERRORS = {
+            '201' => 'order not exist',
+            '204' => 'Some fields from input are null',
+            '205' => 'Cannot find out corresponding key for the user code or user is disabled or user is not activity',
+            '206' => 'Signature from Merchant request is wrong',
+            '207' => 'Param is wrong',
+            '300' => 'Gateway not exist',
+            '301' => 'Wrong account info or currency info',
+            '302' => 'Paycompany not exist',
+            '303' => 'Wrong pay type from merchant',
+            '304' => 'Wallet does not support this payment method',
+            '305' => 'No margin plan for the merchant',
+            '450' => 'Wallet not assign to user',
+            '1'   => 'FAIL'
+          }
+
+          def success?
+            !!@success
+          end
+
+          def errors
+            fail NotImplementedError, "This method must be implemented on the subclass"
+          end
+
+          def error_code_text
+            errors[@response_code]
+          end
+        end
+      end
+
+      class TransactionInterface < Interface
```

Comment:
Not super big, but it'd be nice to see all these classes broken out into separate files